src/nnet3/online-nnet3-decodable.cc

```diff
@@ -0,0 +1,210 @@
+// nnet3/online-nnet3-decodable.cc
+
+// Copyright  2014  Johns Hopkins University (author: Daniel Povey)
+//            2016  Api.ai (Author: Ilya Platonov)
+
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nnet3/online-nnet3-decodable.h"
+#include "nnet3/nnet-utils.h"
+
+namespace kaldi {
+namespace nnet3 {
+
+DecodableNnet3Online::DecodableNnet3Online(
+    const AmNnetSimple &nnet,
+    const TransitionModel &trans_model,
+    const DecodableNnet3OnlineOptions &opts,
+    OnlineFeatureInterface *input_feats):
+    compiler_(nnet.GetNnet(), opts.optimize_config),
+    features_(input_feats),
+    nnet_(nnet),
+    trans_model_(trans_model),
+    opts_(opts),
+    feat_dim_(input_feats->Dim()),
+    num_pdfs_(nnet.GetNnet().OutputDim("output")),
+    begin_frame_(-1) {
+  KALDI_ASSERT(opts_.max_nnet_batch_size > 0);
+  log_priors_ = nnet_.Priors();
+  KALDI_ASSERT((log_priors_.Dim() == 0 || log_priors_.Dim() == trans_model_.NumPdfs()) &&
+               "Priors in neural network must match with transition model (if exist).");
+
+  ComputeSimpleNnetContext(nnet_.GetNnet(), &left_context_, &right_context_);
+  log_priors_.ApplyLog();
+}
+
+
+
+BaseFloat DecodableNnet3Online::LogLikelihood(int32 frame, int32 index) {
+  ComputeForFrame(frame);
+  int32 pdf_id = trans_model_.TransitionIdToPdf(index);
+  KALDI_ASSERT(frame >= begin_frame_ &&
+               frame < begin_frame_ + scaled_loglikes_.NumRows());
+  return scaled_loglikes_(frame - begin_frame_, pdf_id);
+}
+
+
+bool DecodableNnet3Online::IsLastFrame(int32 frame) const {
+  KALDI_ASSERT(false && "Method is not imlemented");
+  return false;
+}
+
+int32 DecodableNnet3Online::NumFramesReady() const {
+  int32 features_ready = features_->NumFramesReady();
+  if (features_ready == 0)
+    return 0;
+  bool input_finished = features_->IsLastFrame(features_ready - 1);
+  if (opts_.pad_input) {
+    // normal case... we'll pad with duplicates of first + last frame to get the
+    // required left and right context.
+    if (input_finished) return subsampling(features_ready);
```

Comment:
Yes, it should start with a capital.  Also the name is not very
explanatory.  Should be called something like NumSubsampledFrames() -- or
just use an inline expression if it's simple.

On Mon, Apr 11, 2016 at 10:40 AM, Ilya Platonova <notifications@github.com>
wrote:

> In src/nnet3/online-nnet3-decodable.cc
> <https://github.com/kaldi-asr/kaldi/pull/666#discussion_r59247523>:
>
> > +
> > +
> > +bool DecodableNnet3Online::IsLastFrame(int32 frame) const {
> > +  KALDI_ASSERT(false && "Method is not imlemented");
> > +  return false;
> > +}
> > +
> > +int32 DecodableNnet3Online::NumFramesReady() const {
> > +  int32 features_ready = features_->NumFramesReady();
> > +  if (features_ready == 0)
> > +    return 0;
> > +  bool input_finished = features_->IsLastFrame(features_ready - 1);
> > +  if (opts_.pad_input) {
> > +    // normal case... we'll pad with duplicates of first + last frame to get the
> > +    // required left and right context.
> > +    if (input_finished) return subsampling(features_ready);
>
> What about subsampling method name not strarting with capital S? :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/666/files/75f8b027a2e6b6b679d26d0b4b667cea631ee6d1#r59247523>
>